jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java

```diff
@@ -527,31 +514,51 @@ void parseAndFillForContent()
 
             assert !_requestBuffer.hasRemaining();
 
+            int filled;
+            boolean compact = true;
+
+            // If the application has retained the content chunks then we must not overwrite content.
             if (_requestBuffer.isRetained())
             {
-                // The application has retained the content chunks,
-                // reacquire the buffer to avoid overwriting the content.
-                releaseRequestBuffer();
-                ensureRequestBuffer();
+                // If there is sufficient space available, we can top up the buffer rather than allocate a new one
+                ByteBuffer backing = _requestBuffer.getByteBuffer();
+                if (BufferUtil.space(backing) >= getHttpConfiguration().getMinInputBufferSpace())
+                {
+                    // do not compact the buffer
+                    compact = false;
+                }
+                else
+                {
+                    // otherwise reacquire the buffer and fill into the new buffer.
+                    releaseRequestBuffer();
+                    ensureRequestBuffer();
+                }
             }
 
-            int filled = fillRequestBuffer();
+            filled = fillRequestBuffer(compact);
+
             if (filled <= 0)
             {
-                releaseRequestBuffer();
+                // Keep the buffer if it is retained
+                if (filled < 0 || !_requestBuffer.isRetained())
```

Comment:
Don't I want to keep around the buffer only if there is enough space?
So here there should be another check against `minInputBufferSpace`.

Reply:
We know there is enough space, because it passed the test above and we just filled 0, so space has not changed.   If it is retained, there is no harm keeping it, as it will not be returned to pool even if we release.  If it isn't retained, then it is empty and unused, so best to release while we wait for more input.